File: 675-CutOffTreesforGolfEvent/675-CutOffTreesforGolfEvent.java
import java.util.*;

class Solution {
    public int cutOffTree(List<List<Integer>> forest) {
        // Special case: Starting position is blocked
        if (forest.get(0).get(0) == 0) return -1;

        // Initialize priority queue to sort trees by height
        PriorityQueue<int[]> pq = new PriorityQueue<>((a, b) -> a[0] - b[0]);
        int m = forest.size();
        int n = forest.get(0).size();

        // Add all trees to the priority queue
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                int height = forest.get(i).get(j);
                if (height > 1) { // Only add trees (height > 1)
                    pq.offer(new int[]{height, i, j});
                }
            }
        }

        // Start from the initial position (0, 0)
        int[] start = new int[]{0, 0};
        int totalSteps = 0;

        // Process trees in order of height
        while (!pq.isEmpty()) {
            int[] nextTree = pq.poll();
            int[] end = new int[]{nextTree[1], nextTree[2]};

            // Perform BFS to find the shortest path from start to end
            int steps = bfs(forest, start, end);
            if (steps == -1) return -1; // If no path exists, return -1

            totalSteps += steps; // Add steps to the total
            start = end; // Move to the next tree
        }

        return totalSteps;
    }

    private int bfs(List<List<Integer>> forest, int[] start, int[] end) {
        int m = forest.size();
        int n = forest.get(0).size();
        int[][] directions = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}}; // Up, Down, Left, Right
        boolean[][] visited = new boolean[m][n]; // Track visited cells
        Queue<int[]> queue = new LinkedList<>();

        queue.offer(start);
        visited[start[0]][start[1]] = true;
        int steps = 0;

        while (!queue.isEmpty()) {
            int size = queue.size();
            for (int i = 0; i < size; i++) {
                int[] current = queue.poll();

                // If we reach the target, return the number of steps
                if (current[0] == end[0] && current[1] == end[1]) {
                    return steps;
                }

                // Explore all four directions
                for (int[] dir : directions) {
                    int newRow = current[0] + dir[0];
                    int newCol = current[1] + dir[1];

                    // Check if the new position is valid and not blocked
                    if (newRow >= 0 && newRow < m && newCol >= 0 && newCol < n &&
                        !visited[newRow][newCol] && forest.get(newRow).get(newCol) >= 1) {
                        queue.offer(new int[]{newRow, newCol});
                        visited[newRow][newCol] = true; // Mark as visited
                    }
                }
            }
            steps++; // Increment steps after processing each level
        }

        // If the target is not reachable, return -1
        return -1;
    }
}

File: 675-CutOffTreesforGolfEvent/675-CutOffTreesforGolfEvent.py
# Last updated: 6/6/2025, 3:27:23 PM
from queue import PriorityQueue
class Solution:
    def cutOffTree(self, forest: List[List[int]]) -> int:
        m, n = len(forest), len(forest[0])
        dirs = {(0,1), (1,0), (0,-1), (-1,0)}
        if forest[0][0] == 0: return -1

        def dst(cur, nxt):
            if cur == nxt: return 0
            i, j = cur[1], cur[2]
            target_i, target_j = nxt[1], nxt[2]
            # print("i,j: ",i,"; ",j)
            q = deque([(i, j)])
            visited = set([(i, j)])
            # print("visited: ",visited)
            steps = 0
            while q:
                # print("q: ", q)
                length = len(q)
                # print(length)
                steps += 1
                for _ in range(length):
                    c = q.popleft()
                    # print("steps: ",steps, "; c: ",c, ",target_i, target_j:",target_i,"; ", target_j)
                    for dir in dirs:
                        nxt_i, nxt_j = c[0] + dir[0],c[1] + dir[1]
                        
                        if 0 <= nxt_i < m and 0 <= nxt_j < n:
                            if nxt_i == target_i and nxt_j == target_j:
                                return steps
                            if (nxt_i, nxt_j) not in visited and forest[nxt_i][nxt_j] > 0:
                                visited.add((nxt_i, nxt_j))
                                q.append((nxt_i, nxt_j))
            return -1


        pq = PriorityQueue()
        for i in range(m):
            for j in range(n):
                if forest[i][j] <= 1:
                    continue
                pq.put((forest[i][j], i, j))
        cur = (forest[0][0],0,0)
        # print(pq)
        # print("cur: ",cur)
        res = 0
        while not pq.empty():
            nxt = pq.get()
            # print("nxt: ",nxt)
            d = dst(cur, nxt)
            if d < 0:
                return -1
            res += d
            cur = nxt
        return res
